### PR TITLE
Feat!: Extend SQL parser with designated blocks for jinja code

### PR DESCRIPTION
**Pull request by izeigerman**
This update isolates Jinja expressions into designated blocks to simplify SQL parsing and capture user's intention more precisely. 

Model definition example:
```sql
MODEL (
  name sqlmesh_example.full_model
);

JINJA_STATEMENT_BEGIN;
{{ pre_hook() }}
JINJA_END;

JINJA_QUERY_BEGIN;
SELECT {{ 1 + 1 }};
JINJA_END;

JINJA_STATEMENT_BEGIN;
{{ post_hook() }}
JINJA_END;
```

### CODE REVIEW

File: sqlmesh/core/dialect.py [hunk 1]
```diff
@@ -404,37 +455,51 @@ def parse(sql: str, default_dialect: t.Optional[str] = None) -> t.List[exp.Expre
...
-        if is_jinja:
+    for chunk, chunk_type in chunks:
+        if chunk_type == ChunkType.SQL:
+            for expression in dialect.parser().parse(chunk, sql):
+                if expression:
+                    expressions.append(expression)
+        else:
             start, *_, end = chunk
             segment = sql[start.start : end.end + 2]
             variables = [
```
Comment (tobymao) [inline sqlmesh/core/dialect.py:495]: can't you delete this section and get rid of expressions?

File: sqlmesh/migrations/v0012_update_jinja_expressions.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+def _wrap_query(sql: str) -> str:
+    return f"JINJA_QUERY_BEGIN;\n{sql}\nJINJA_END;"
+
+
+def _wrap_statement(sql: str) -> str:
+    return f"JINJA_STATEMENT_BEGIN;\n{sql}\nJINJA_END;"
+
+
+JINJA_REGEX = r"({{|{%)"
```
Comment (tobymao) [inline sqlmesh/migrations/v0012_update_jinja_expressions.py:81]: re.compile(...)

File: sqlmesh/core/dialect.py [hunk 3]
```diff
@@ -387,6 +397,47 @@ def text_diff(
...
+            tokens[pos].text.upper() == command.upper()
+            and tokens[pos + 1].token_type == TokenType.SEMICOLON
+        )
+    except IndexError:
+        return False
+
+
+JINJA_QUERY_BEGIN = "JINJA_QUERY_BEGIN"
+JINJA_STATEMENT_BEGIN = "JINJA_STATEMENT_BEGIN"
+JINJA_END = "JINJA_END"
```
[the remark this answers is not in the document]
Comment (tobymao) [reply sqlmesh/core/dialect.py:412]: hard to create tokens because they are enums

[the remark this answers is not in the document]
Comment (GeorgeSittas) [reply sqlmesh/core/dialect.py:412]: Would it be possible to inject new enum values in `TokenType` through metaprogramming to achieve this?

[the remark this answers is not in the document]
Comment (tobymao) [reply sqlmesh/core/dialect.py:412]: https://stackoverflow.com/questions/28126314/adding-members-to-python-enums

**Pull request merged**